Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/internal/GarlandHeckbert_core.h

```diff
@@ -0,0 +1,313 @@
+  typedef typename Geom_traits::Plane_3                                         Plane_3;
+  typedef typename Geom_traits::Vector_3                                        Vector_3;
+
+  typedef typename GarlandHeckbert_matrix_type<TM_, GT_>::type                  Matrix4x4;
+  typedef Eigen::Matrix<FT, 1, 4>                                               Row4;
+  typedef Eigen::Matrix<FT, 4, 1>                                               Col4;
+
+  static Col4 point_to_homogenous_column(const Point& pt)
+  {
+    return Col4(pt.x(), pt.y(), pt.z(), FT(1));
+  }
+
+  /**
+  * Combines two Q matrices.
+  * It is simply the addition of two matrices
+  */
+  inline static Matrix4x4 combine_matrices(const Matrix4x4& aFirst, const Matrix4x4& aSecond)
+  {
+    return aFirst + aSecond;
+  }
+
+  /**
+  * Returns `true` if the target of h is a discontinuity vertex in tmesh
+  *
+  * Currently, only checks if vertex belongs to a border.
+  */
+  static bool is_discontinuity_vertex(const halfedge_descriptor h, const Triangle_mesh& tmesh)
+  {
+    if(is_border(target(h, tmesh), tmesh))
+      return true;
+
+    return false;
+  }
+
+  static bool is_discontinuity_edge(const halfedge_descriptor h, const Triangle_mesh& tmesh)
+  {
+    return is_border_edge(h, tmesh);
+  }
+
+  /*
+  * fundamental error quadric for the target vertex of h in tmesh
+  * Unused, but leaving it as it might still be useful somehow
+  */
+  static Matrix4x4 fundamental_error_quadric(const halfedge_descriptor h,
+                                             const Triangle_mesh& tmesh,
+                                             const Vertex_point_pmap& vpm,
+                                             const Geom_traits& gt,
+                                             const FT discontinuity_multiplier = FT(100))
+  {
+    Matrix4x4 quadric = Matrix4x4::Zero();
+
+    const vertex_descriptor target_vd = target(h, tmesh);
+    const Vector_3 target_vertex_vector = gt.construct_vector_3_object()(CGAL::ORIGIN, get(vpm, target_vd));
+
+    // const bool discontinuity_vertex = is_discontinuity_vertex(h, tmesh);
+
+    for(const halfedge_descriptor hd : halfedges_around_target(target_vd, tmesh))
+    {
+      const face_descriptor fd = face(hd, tmesh);
+      if(fd == GraphTraits::null_face())
+        continue;
+
+      Plane_3 plane(get(vpm, source(hd, tmesh)),
+                    get(vpm, target(hd, tmesh)),
+                    get(vpm, target(next(hd, tmesh), tmesh)));
+
+      Row4 plane_mtr;
+      const FT norm = sqrt(CGAL::square(plane.a()) +
+                           CGAL::square(plane.b()) +
+                           CGAL::square(plane.c()));
+      const FT den = FT(1) / norm;
+
+      plane_mtr << den * plane.a(),
+                   den * plane.b(),
+                   den * plane.c(),
+                   den * plane.d();
+      quadric += plane_mtr.transpose() * plane_mtr;
+
+      if(is_discontinuity_edge(hd, tmesh))
```

Comment:
When I have twice the same code for two opposite halfedges, I usually do a for-loop(i=0;i<2) and update hd to its opposite